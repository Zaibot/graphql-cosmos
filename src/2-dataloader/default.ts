import { createDataLoader, DataLoaderHandler } from './loader'
import { defaultOnDataLoaderResolve } from './resolver'

export const defaultDataLoader = (): DataLoaderHandler => {
  const loader = createDataLoader({})
  return loader
}
