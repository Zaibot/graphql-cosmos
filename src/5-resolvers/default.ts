import { GraphQLCosmosFieldResolver } from './resolver'
import { SourceDescriptor } from './x-descriptors'

export const defaultCosmosFieldResolver: GraphQLCosmosFieldResolver = async (parent, _args, context, info) => {
  const field = context.dataSources.graphqlCosmos.meta.requireField(info.parentType.name, info.fieldName)
  const column = field.ours ?? field.fieldname
  const source = SourceDescriptor.getDescriptor(parent)
  if (source?.kind === `Single`) {
    const value = await context.dataSources.graphqlCosmos.load(source, column)
    return value
  } else {
    const value = Object(parent)[column]
    return value
  }
}

// export const defaultCosmosWhereFieldResolver: GraphQLCosmosFieldResolver = (parent, _args, context, info) => {
//   const field = context.dataSources.graphqlCosmos.meta.requireField(info.parentType.name, info.fieldName)
//   const column = field.whereOurs ?? field.ours ?? field.fieldname
//   const source = SourceDescriptor.getDescriptor(parent)
//   if (source?.kind === `Single`) {
//     return context.dataSources.graphqlCosmos.load(source, column)
//   } else {
//     const value = Object(parent)[column]
//     console.log(`EMBEDDED ${info.parentType}.${info.fieldName} = ${value}`)
//     return value
//   }
// }

// export const defaultCosmosSortFieldResolver: GraphQLCosmosFieldResolver = (parent, _args, context, info) => {
//   const field = context.dataSources.graphqlCosmos.meta.requireField(info.parentType.name, info.fieldName)
//   const column = field.sortOurs ?? field.ours ?? field.fieldname
//   const source = SourceDescriptor.getDescriptor(parent)
//   if (source?.kind === `Single`) {
//     return context.dataSources.graphqlCosmos.load(source, column)
//   } else {
//     const value = Object(parent)[column]
//     return value
//   }
// }
