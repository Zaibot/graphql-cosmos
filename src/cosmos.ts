import { SqlParameter } from '@azure/cosmos';
import { GraphQLCosmosRequest } from './context';
import { SqlBuilder } from './sql';

export const defaultOnQuery = ({ client, database, container, query, parameters }: GraphQLCosmosRequest) =>
    client.database(database).container(container).items.query({ query, parameters }).fetchAll();

export function createSqlQuery(
    whereExpressions: Array<{ property: string; operation: string; value: unknown; parameter: string }>,
    sortExpressions: Array<{ property: string; direction: string }>,
) {
    const fromAlias = `c`;
    const expressions = whereExpressions.map((expr) => {
        const operationMap: Record<string, string> = {
            '': `${fromAlias}.${expr.property} = ${expr.parameter}`,
            eq: `${fromAlias}.${expr.property} = ${expr.parameter}`,
            neq: `${fromAlias}.${expr.property} != ${expr.parameter}`,
            gt: `${fromAlias}.${expr.property} > ${expr.parameter}`,
            gte: `${fromAlias}.${expr.property} >= ${expr.parameter}`,
            lt: `${fromAlias}.${expr.property} < ${expr.parameter}`,
            lte: `${fromAlias}.${expr.property} <= ${expr.parameter}`,
            in: `ARRAY_CONTAINS(${expr.parameter}, ${fromAlias}.${expr.property})`,
            nin: `NOT ARRAY_CONTAINS(${expr.parameter}, ${fromAlias}.${expr.property})`,
            contains: `CONTAINS(${fromAlias}.${expr.property}, ${expr.parameter})`,
            ncontains: `NOT CONTAINS(${fromAlias}.${expr.property}, ${expr.parameter})`,
        };
        const sql: string = operationMap[expr.operation];
        const parameter: SqlParameter = { name: expr.parameter, value: expr.value as any };
        if (sql) {
            return { sql, parameter };
        } else {
            throw Error(`unknown expression in ${JSON.stringify(whereExpressions)}`);
        }
    });

    const sql = new SqlBuilder(fromAlias);
    for (const expr of expressions) {
        if (expr) {
            sql.where(expr.sql);
        }
    }

    for (const { property, direction } of sortExpressions) {
        if (direction === `ASC`) {
            sql.orderBy(`${fromAlias}.${property}`, `ASC`);
        } else if (direction === `DESC`) {
            sql.orderBy(`${fromAlias}.${property}`, `DESC`);
        } else {
            throw Error(`sort direction of ${property} must be ASC or DESC`);
        }
    }

    sql.orderBy(`${fromAlias}.id`, `ASC`);

    const parameters = expressions.map((x) => x?.parameter).filter((x): x is SqlParameter => !!x);

    return { sql, parameters };
}
