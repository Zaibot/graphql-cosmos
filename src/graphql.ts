import { GraphQLInputFieldConfig, GraphQLSchema, GraphQLInputType, GraphQLInputObjectType, GraphQLField } from 'graphql';

export const createOrGetWhereType = (name: string, properties: Record<string, GraphQLInputFieldConfig>, schema: GraphQLSchema) => {
    if (schema.getTypeMap()[name]) {
        return schema.getTypeMap()[name] as GraphQLInputType;
    }

    const filterType = new GraphQLInputObjectType({
        name,
        description: undefined,
        extensions: undefined,
        fields: properties,
        astNode: undefined,
        extensionASTNodes: undefined,
    });

    // HACK
    schema.getTypeMap()[name] = filterType;

    return filterType;
};

export const addFieldArgument = (field: GraphQLField<any, any, any>, name: string, type: GraphQLInputType) => {
    field.args.push({
        name,
        type,
        astNode: undefined,
        extensions: undefined,
        description: undefined,
        defaultValue: undefined,
    });
};
