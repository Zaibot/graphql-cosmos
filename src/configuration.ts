import { CosmosClient, FeedOptions, FeedResponse } from '@azure/cosmos'
import { DataLoaderHandler } from './dataloader/loader'
import { CosmosRequest } from './intermediate/model'
import { SqlBuilder } from './sql/builder'

export interface GraphQLCosmosContext {
  directives: GraphQLDirectivesContext
}

export type CosmosQueryHandler = (request: GraphQLCosmosRequest) => Promise<FeedResponse<any>>

export interface GraphQLDirectivesContext {
  cosmos: {
    database: string
    client: CosmosClient
    dataloader?: DataLoaderHandler<GraphQLCosmosContext>
    /** default: defaultOnInit */
    onInit?: (request: CosmosRequest, init: GraphQLCosmosInitRequest) => void
    onBeforeQuery?: (request: GraphQLCosmosInitRequest) => void
    /** default: defaultOnQuery */
    onQuery?: CosmosQueryHandler
  }
}

export interface GraphQLCosmosInitRequest {
  client: CosmosClient
  request: CosmosRequest
  database: string
  container: string
  query?: SqlBuilder
  parameters?: Array<{ name: string; value: any }>
  options?: FeedOptions
}

export interface GraphQLCosmosRequest {
  init?: GraphQLCosmosInitRequest
  client: CosmosClient
  database: string
  container: string
  query: string
  parameters: Array<{ name: string; value: any }>
  options?: FeedOptions
}
