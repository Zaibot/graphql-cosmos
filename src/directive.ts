import { DirectiveLocation, getNullableType, GraphQLDirective, GraphQLField, GraphQLInterfaceType, GraphQLObjectType, GraphQLScalarType, GraphQLSchema, GraphQLString, isListType, isObjectType, isScalarType } from 'graphql';
import { SchemaDirectiveVisitor } from 'graphql-tools';
import { GraphQLCosmosContext, GraphQLCosmosInitRequest, GraphQLCosmosRequest } from './context';
import { createSqlQuery, defaultOnQuery } from './cosmos';
import { addFieldArgument, createOrGetWhereType } from './graphql';

export class CosmosDirective extends SchemaDirectiveVisitor {
    static getDirectiveDeclaration(directiveName: string, schema: GraphQLSchema) {
        const previousDirective = schema.getDirective(directiveName);
        if (previousDirective) {
            return previousDirective;
        }

        return new GraphQLDirective({
            name: directiveName,
            locations: [DirectiveLocation.OBJECT, DirectiveLocation.FIELD_DEFINITION],
            args: {
                container: {
                    type: GraphQLString,
                },
                ours: {
                    type: GraphQLString,
                },
                theirs: {
                    type: GraphQLString,
                },
            },
        });
    }

    get argContainer() {
        return this.args.container as string | undefined;
    }

    get argOurs() {
        return this.args.ours as string | undefined;
    }

    get argTheirs() {
        return this.args.theirs as string | undefined;
    }

    visitFieldDefinition(
        field: GraphQLField<any, any>,
        _details: {
            objectType: GraphQLObjectType | GraphQLInterfaceType;
        },
    ): GraphQLField<any, any> | void | null {
        const operations = [``, `_eq`, `_neq`, `_gt`, `_gte`, `_lt`, `_lte`, `_in`, `_nin`, `_contain`, `_ncontain`];

        if (isListType(field.type)) {
            const otype = field.type.ofType;
            if (isObjectType(otype)) {
                const ours = this.argOurs;
                const theirs = this.argTheirs;
                const container = this.argContainer;
                if (!container) throw Error(`requires container argument`);

                //
                // Get scalar fields
                const filterable = Object.entries(otype.getFields())
                    .filter(([_, f]) => isScalarType(getNullableType(f.type)))
                    .map(([name, tmpfield]) => ({ name, tmpfield, scalar: getNullableType(tmpfield.type) as GraphQLScalarType }));

                //
                // Replace resolver
                field.resolve = async (source, args, context: GraphQLCosmosContext, info) => {
                    if (ours || theirs) {
                        const ourValueOrList = source[ours ?? 'id'];
                        if (Array.isArray(ourValueOrList)) {
                            const whereOurs = `${theirs ?? 'id'}_in`;
                            if (whereOurs in args) throw Error(`argument contains conflicting filter on ${whereOurs}`);
                            return await this.collectionResolver(otype, { where: { ...args, [whereOurs]: ourValueOrList } }, context, container);
                        } else {
                            const whereOurs = `${theirs ?? 'id'}_eq`;
                            if (whereOurs in args) throw Error(`argument contains conflicting filter on ${whereOurs}`);
                            return await this.collectionResolver(otype, { where: { ...args, [whereOurs]: ourValueOrList } }, context, container);
                        }
                    } else {
                        return await this.collectionResolver(otype, args, context, container);
                    }
                };

                //
                // Generate where type holder filter input
                const whereFields = Object.fromEntries(
                    filterable.flatMap((field) =>
                        operations
                            .map((operation) => ({ ...field, operation }))
                            .map(({ name, operation, scalar }) => [`${name}${operation}`, { type: scalar, extensions: { __operation: operation } }]),
                    ),
                );
                const filterType = createOrGetWhereType(`${field.type.ofType.name}Where`, whereFields, this.schema);

                addFieldArgument(field, `where`, filterType);
            }
        }
        return field;
    }

    private async collectionResolver(type: GraphQLObjectType<any, any>, args: Record<string, any>, context: GraphQLCosmosContext, container: string) {
        const { where = {} } = args;
        const { cosmos } = context.directives;
        const { onBeforeQuery, onQuery = defaultOnQuery } = cosmos;

        //
        // Parse GraphQL where fields
        const whereInputExpressions = Object.entries(where).map(([whereField, value]) => {
            const [property, operation = ``] = whereField.split(`_`);
            return { property, operation, value, parameter: `@${whereField}` };
        });

        //
        // Construct query
        const { sql, parameters } = createSqlQuery(whereInputExpressions);

        //
        // Prepare CosmosDB query
        const init: GraphQLCosmosInitRequest = {
            client: cosmos.client,
            database: cosmos.database,
            container,
            query: sql,
            parameters,
        };
        onBeforeQuery?.(init);

        //
        // Send CosmosDB query
        const request: GraphQLCosmosRequest = {
            client: init.client,
            database: init.database,
            container: init.container,
            query: init.query.toSql(),
            parameters: init.parameters,
        };
        const response = await onQuery(request);

        return response.resources.map((x) => ({ __typename: type.name, ...x }));
    }

    visitObject(object: GraphQLObjectType) {
        // console.log(`visitObject(${object.name})`, this.args);
    }
}
