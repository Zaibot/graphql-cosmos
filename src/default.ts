import { GraphQLCosmosInitRequest, GraphQLCosmosRequest } from './configuration'
import { createDataLoader, DataLoaderHandler } from './dataloader/loader'
import { defaultOnDataLoaderResolve } from './dataloader/resolver'
import { CosmosRequest } from './intermediate/model'
import { convertToSql } from './sql'

export const defaultOnQuery = ({ client, database, container, query, parameters, options }: GraphQLCosmosRequest) =>
  client.database(database).container(container).items.query({ query, parameters }, options).fetchNext()

export const defaultOnInit = (request: CosmosRequest, init: GraphQLCosmosInitRequest) => {
  const sql = convertToSql(request)
  init.query = sql.sql
  init.parameters = sql.parameters
  init.options ??= {}
  init.options.continuationToken = request.cursor
}

export const defaultDataLoader = (): DataLoaderHandler => {
  const loader = createDataLoader({ resolve: defaultOnDataLoaderResolve })
  return loader
}
