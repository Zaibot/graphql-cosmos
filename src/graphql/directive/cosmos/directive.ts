import * as GraphQL from 'graphql'
import {
  DirectiveLocation,
  DirectiveNode,
  getDirectiveValues,
  getNamedType,
  getNullableType,
  GraphQLSchema,
  isListType,
  isObjectType,
  isScalarType,
} from 'graphql'
import { SchemaDirectiveVisitor } from 'graphql-tools'
import { debugHooks as debugHook } from '../../../debug'
import { withErrorMiddleware } from '../../../error'
import { addFieldArgument, createOrGetPageType } from '../../internal/schema'
import { SortDirective } from '../sort/directive'
import { inputSort } from '../sort/input'
import { WhereDirective } from '../where/directive'
import { inputWhere } from '../where/input'
import {
  resolveManyOurs,
  resolveManyOursWithoutContainer,
  resolveManyTheirs,
  resolveOneOurs,
  resolveOneTheirs,
  resolveRootQuery,
} from './resolvers'

export class CosmosDirective extends SchemaDirectiveVisitor {
  static getDirectiveDeclaration(directiveName: string, schema: GraphQL.GraphQLSchema) {
    const previousDirective = schema.getDirective(directiveName)
    if (previousDirective) {
      return previousDirective
    }

    return new GraphQL.GraphQLDirective({
      name: directiveName,
      locations: [DirectiveLocation.FIELD_DEFINITION],
      args: {
        container: {
          type: GraphQL.GraphQLString,
        },
        ours: {
          type: GraphQL.GraphQLString,
        },
        theirs: {
          type: GraphQL.GraphQLString,
        },
      },
    })
  }

  static getContainer(
    directiveName: string,
    schema: GraphQLSchema,
    node: { readonly directives?: readonly DirectiveNode[] | undefined }
  ) {
    const directive = WhereDirective.getDirectiveDeclaration(directiveName, schema)
    const values = getDirectiveValues(directive, node)
    const raw = values?.container as string | undefined
    return raw
  }
  static getOurs(
    directiveName: string,
    schema: GraphQLSchema,
    node: { readonly directives?: readonly DirectiveNode[] | undefined }
  ) {
    const directive = WhereDirective.getDirectiveDeclaration(directiveName, schema)
    const values = getDirectiveValues(directive, node)
    const raw = values?.orus as string | undefined
    return raw
  }
  get argContainer() {
    return this.args.container as string | undefined
  }

  get argOurs() {
    return this.args.ours as string | undefined
  }

  get argTheirs() {
    return this.args.theirs as string | undefined
  }

  visitFieldDefinition(
    fieldType: GraphQL.GraphQLField<any, any>,
    details: {
      objectType: GraphQL.GraphQLObjectType | GraphQL.GraphQLInterfaceType
    }
  ): GraphQL.GraphQLField<any, any> | void | null {
    if (!GraphQL.isObjectType(details.objectType)) {
      throw Error(`must be an ObjectType: ${details.objectType.name}`)
    }

    const directiveNameWhere = `where`
    const directiveNameSort = `sort`

    const ours = this.argOurs
    const theirs = this.argTheirs
    const theirContainer = this.argContainer

    const returnType = fieldType.type
    const returnTypeCore = getNamedType(returnType)
    const returnTypeMany = isListType(getNullableType(returnType)) ? returnType : undefined

    if (theirContainer) {
      if (isObjectType(returnTypeCore)) {
        const filterableScalar = Object.entries(returnTypeCore.getFields())
          .map(([name, field]) => ({
            name,
            field,
            scalar: getNullableType(field.type) as GraphQL.GraphQLScalarType,
            fieldIsArray: Boolean(isListType(field.type)),
          }))
          .filter(
            ({ field }) => isScalarType(getNullableType(field.type)) || GraphQL.isEnumType(getNullableType(field.type))
          )
          .map(({ field, name, scalar, fieldIsArray }) => ({
            name,
            scalar,
            fieldIsArray,
            operations: WhereDirective.getOp(directiveNameWhere, this.schema, field.astNode!) ?? [],
          }))
          .filter((x) => x.operations.length > 0)
        if (filterableScalar.length > 0) {
          const filterType = inputWhere(returnTypeCore.name, filterableScalar, this.schema)
          addFieldArgument(fieldType, `where`, filterType)
        }

        const sortableScalar = Object.entries(returnTypeCore.getFields())
          .map(([name, field]) => ({ name, field, scalar: getNullableType(returnType) as GraphQL.GraphQLScalarType }))
          .filter(
            ({ field }) => isScalarType(getNullableType(field.type)) || GraphQL.isEnumType(getNullableType(field.type))
          )
          .filter(({ field }) => SortDirective.has(directiveNameSort, field.astNode ?? {}))
        if (sortableScalar.length > 0) {
          const sortType = inputSort(returnTypeCore.name, sortableScalar, this.schema)
          addFieldArgument(fieldType, `sort`, sortType)
        }

        //
        // Override resolvers per relation type
        if (returnTypeMany && theirs) {
          debugHook?.onResolverSet?.({
            resolver: `resolveManyTheirs`,
            objectType: details.objectType,
            fieldType,
          })
          addFieldArgument(fieldType, `cursor`, GraphQL.GraphQLString)
          fieldType.type = new GraphQL.GraphQLNonNull(wrapOutputWithPagination(fieldType.type, this.schema))
          fieldType.resolve = withErrorMiddleware(
            `resolveManyTheirs`,
            resolveManyTheirs(theirContainer, ours, theirs, fieldType)
          )
        } else if (returnTypeMany && ours) {
          debugHook?.onResolverSet?.({
            resolver: `resolveManyOurs`,
            objectType: details.objectType,
            fieldType,
          })
          addFieldArgument(fieldType, `cursor`, GraphQL.GraphQLString)
          fieldType.type = new GraphQL.GraphQLNonNull(wrapOutputWithPagination(fieldType.type, this.schema))
          fieldType.resolve = withErrorMiddleware(
            `resolveManyOurs`,
            resolveManyOurs(theirContainer, ours, theirs, fieldType)
          )
        } else if (returnTypeMany) {
          debugHook?.onResolverSet?.({
            resolver: `resolveRootQuery`,
            objectType: details.objectType,
            fieldType,
          })
          addFieldArgument(fieldType, `cursor`, GraphQL.GraphQLString)
          fieldType.type = new GraphQL.GraphQLNonNull(wrapOutputWithPagination(fieldType.type, this.schema))
          fieldType.resolve = withErrorMiddleware(`resolveRootQuery`, resolveRootQuery(theirContainer, fieldType))
        } else if (ours) {
          debugHook?.onResolverSet?.({
            resolver: `resolveOneOurs`,
            objectType: details.objectType,
            fieldType,
          })
          fieldType.resolve = withErrorMiddleware(`resolveOneOurs`, resolveOneOurs(ours, fieldType))
        } else if (theirs) {
          debugHook?.onResolverSet?.({
            resolver: `resolveOneTheirs`,
            objectType: details.objectType,
            fieldType,
          })
          fieldType.resolve = withErrorMiddleware(`resolveOneTheirs`, resolveOneTheirs(theirContainer, ours, fieldType))
        }
      }
    } else if (ours) {
      if (returnTypeMany) {
        // No container, ours specifies id field to be created as reference
        debugHook?.onResolverSet?.({
          resolver: `resolveManyOursWithoutContainer`,
          objectType: details.objectType,
          fieldType,
        })
        fieldType.resolve = withErrorMiddleware(
          `resolveManyOursWithoutContainer`,
          resolveManyOursWithoutContainer(ours, fieldType)
        )
      } else {
        // No container, ours specifies id field to be created as reference
        debugHook?.onResolverSet?.({
          resolver: `resolveOneOurs`,
          objectType: details.objectType,
          fieldType,
        })
        fieldType.resolve = withErrorMiddleware(`resolveOneOurs`, resolveOneOurs(ours, fieldType))
      }
    }

    return fieldType
  }
}

const wrapOutputWithPagination = (type: GraphQL.GraphQLOutputType, schema: GraphQL.GraphQLSchema) => {
  return createOrGetPageType(
    `${getNamedType(type).name}Page`,
    {
      nextCursor: { type: GraphQL.GraphQLString },
      total: { type: new GraphQL.GraphQLNonNull(GraphQL.GraphQLInt) },
      page: { type },
    },
    schema
  )
}
