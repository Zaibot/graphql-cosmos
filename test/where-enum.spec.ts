import { FeedResponse } from '@azure/cosmos'
import { execute, GraphQLSchema, validate, validateSchema } from 'graphql'
import gql from 'graphql-tag'
import { makeExecutableSchema } from 'graphql-tools'
import { GraphQLCosmosContext, GraphQLCosmosRequest } from '../src/configuration'
import { defaultDataLoader } from '../src/default'
import { schema } from '../src/graphql/directive/schema'

const dummyTypeDefs = gql`
  type Query {
    entities: [Entity!]! @cosmos(container: "Entities")
  }

  type Entity {
    id: ID!
    status: Status! @where(op: "eq neq")
  }

  enum Status {
    OPEN
    CLOSE
  }
`

const onCosmosQuery = async (request: GraphQLCosmosRequest): Promise<FeedResponse<unknown>> => {
  const { container, query, parameters } = request
  const key = parameters.length ? `${query} (${parameters.map((x) => `${x.name}=${x.value}`).toString()})` : query

  const responses: Record<string, Record<string, unknown[]>> = {
    Entities: {
      'SELECT c.id FROM c ORDER BY c.id': [{ id: `1` }, { id: `2` }],
      'SELECT r.id, r.status FROM r WHERE ARRAY_CONTAINS(@batch, r.id) (@batch=1)': [
        {
          __typename: 'Entity',
          id: `1`,
          status: `OPEN`,
        },
      ],
      'SELECT r.id, r.status FROM r WHERE ARRAY_CONTAINS(@batch, r.id) (@batch=2)': [
        {
          __typename: 'Entity',
          id: `2`,
          status: `CLOSE`,
        },
      ],
      'SELECT c.id FROM c WHERE c.status = @status_eq ORDER BY c.id (@status_eq=OPEN)': [{ id: `1` }],
      'SELECT c.id FROM c WHERE c.status != @status_neq ORDER BY c.id (@status_neq=OPEN)': [{ id: `2` }],
    },
  }

  const result = responses[container]?.[key]
  if (result) {
    return { resources: result } as any
  } else {
    throw Error(`Unhandled: ${container} ${key}`)
  }
}

describe(`Where`, () => {
  let context: GraphQLCosmosContext
  let dummy: GraphQLSchema

  beforeEach(() => {
    context = {
      directives: {
        cosmos: {
          database: null as any,
          client: null as any,
          onQuery: onCosmosQuery,
          dataloader: defaultDataLoader(onCosmosQuery),
        },
      },
    }

    dummy = makeExecutableSchema({
      typeDefs: [schema.typeDefs, dummyTypeDefs],
      schemaDirectives: {
        ...schema.schemaDirectives,
      },
    })

    expect(validateSchema(dummy)).toHaveLength(0)
  })

  it(`should retrieve all`, async () => {
    const query = gql`
      query {
        entities {
          page {
            __typename
            id
            status
          }
        }
      }
    `
    expect(validate(dummy, query)).toHaveLength(0)

    const result = await execute(dummy, query, undefined, context)
    result.errors?.forEach((e) => console.error(e.path.join(`/`), e.stack))

    expect(result).toEqual({
      data: {
        entities: {
          page: [
            {
              __typename: 'Entity',
              id: `1`,
              status: `OPEN`,
            },
            {
              __typename: 'Entity',
              id: `2`,
              status: `CLOSE`,
            },
          ],
        },
      },
    })
  })

  it(`should retrieve all open`, async () => {
    const query = gql`
      query {
        entities(where: { status_eq: OPEN }) {
          page {
            __typename
            id
            status
          }
        }
      }
    `
    expect(validate(dummy, query)).toHaveLength(0)

    const result = await execute(dummy, query, undefined, context)
    result.errors?.forEach((e) => console.error(e.path.join(`/`), e.stack))

    expect(result).toEqual({ data: { entities: { page: [{ __typename: 'Entity', id: `1`, status: `OPEN` }] } } })
  })

  it(`should retrieve all close`, async () => {
    const query = gql`
      query {
        entities(where: { status_neq: OPEN }) {
          page {
            __typename
            id
            status
          }
        }
      }
    `
    expect(validate(dummy, query)).toHaveLength(0)

    const result = await execute(dummy, query, undefined, context)
    result.errors?.forEach((e) => console.error(e.path.join(`/`), e.stack))

    expect(result).toEqual({ data: { entities: { page: [{ __typename: 'Entity', id: `2`, status: `CLOSE` }] } } })
  })
})
