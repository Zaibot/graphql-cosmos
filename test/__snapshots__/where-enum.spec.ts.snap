// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Where expects meta schema to remain the same 1`] = `
Object {
  "types": Array [
    Object {
      "container": "Entities",
      "cosmos": true,
      "database": "Test",
      "external": false,
      "fields": Array [
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "id",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "status",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "Status",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "eq",
            "neq",
            "in",
            "nin",
          ],
          "whereOurs": null,
        },
      ],
      "filterable": true,
      "sortable": false,
      "typename": "Entity",
    },
    Object {
      "container": null,
      "cosmos": false,
      "database": null,
      "external": false,
      "fields": Array [
        Object {
          "container": "Entities",
          "cosmos": true,
          "database": "Test",
          "fieldname": "entities",
          "ours": null,
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Entity",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Query",
    },
  ],
}
`;

exports[`Where expects schema to remain the same 1`] = `
"schema {
  query: Query
}
type Query {
  entities(where: EntityWhere, cursor: String, limit: Int): EntityPage!
}
type Entity {
  id: ID!
  status: Status!
}
enum Status {
  NONE
  OPEN
  CLOSE
}
type EntityPage {
  total: Int!
  cursor: String
  nextCursor: String
  page: [Entity]!
}
input EntityWhere {
  and: [EntityWhere!]
  or: [EntityWhere!]
  status_eq: Status
  status_neq: Status
  status_in: [Status]
  status_nin: [Status]
}"
`;
